File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int[] dp = new int[obstacleGrid[0].length];
        for(int i = 0; i < dp.length; ++i) {
            if(obstacleGrid[0][i] == 1)
                dp[i] = 0;
            else {
                if(i == 0)
                    dp[i] = 1;
                else
                    dp[i] = dp[i-1];
            }
        }
        for(int y = 1; y < obstacleGrid.length; ++y) {
            if(obstacleGrid[y][0] == 1)
                dp[0] = 0;
            for(int x = 1; x < obstacleGrid[0].length; ++x) {
                if(obstacleGrid[y][x] == 1)
                    dp[x] = 0;
                else
                    dp[x] += dp[x-1];
            }
        }
        return dp[dp.length - 1];
    }
}

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        vector<int> dp = obstacleGrid[0];
        dp[0] = 1;
        for(int y = 0; y < obstacleGrid.size(); ++y) {
            if(obstacleGrid[y][0] == 1)
                dp[0] = 0;
            for(int x = 1; x < obstacleGrid[0].size(); ++x)
                dp[x] = (dp[x-1] + dp[x]) * (1 - obstacleGrid[y][x]);
        }
        return dp[dp.size()-1];
    }
};